go/indexer/service.go

```diff
@@ -0,0 +1,309 @@
+	gwei := new(big.Float).SetInt(wei)
+	gwei.Mul(gwei, weiToGwei)
+	gwei64, _ := gwei.Float64()
+	return gwei64
+}
+
+// Driver is an interface for indexing deposits from l1.
+type Driver interface {
+	// Name is an identifier used to prefix logs for a particular service.
+	Name() string
+
+	// Metrics returns the subservice telemetry object.
+	Metrics() *metrics.Metrics
+}
+
+type ServiceConfig struct {
+	Context            context.Context
+	L1Client           *ethclient.Client
+	ChainID            *big.Int
+	CTCAddr            common.Address
+	ConfDepth          uint64
+	MaxHeaderBatchSize uint64
+	StartBlockNumber   uint64
+	StartBlockHash     string
+	DB                 *Database
+}
+
+type Service struct {
+	cfg    ServiceConfig
+	ctx    context.Context
+	cancel func()
+
+	contract       *ctc.CanonicalTransactionChainFilterer
+	backend        Backend
+	headerSelector HeaderSelector
+
+	metrics *metrics.Metrics
+
+	wg sync.WaitGroup
+}
+
+func NewService(cfg ServiceConfig) (*Service, error) {
+	ctx, cancel := context.WithCancel(cfg.Context)
+
+	address := cfg.CTCAddr
+	contract, err := ctc.NewCanonicalTransactionChainFilterer(address, cfg.L1Client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle restart logic
+
+	log.Info("Creating CTC Indexer")
+
+	chainID, err := cfg.L1Client.ChainID(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.ChainID != nil {
+		if cfg.ChainID.Cmp(chainID) != 0 {
+			return nil, fmt.Errorf("%w: configured with %d and got %d",
+				errWrongChainID, cfg.ChainID, chainID)
+		}
+	} else {
+		cfg.ChainID = chainID
+	}
+
+	confirmedHeaderSelector, err := NewConfirmedHeaderSelector(HeaderSelectorConfig{
+		ConfDepth:    cfg.ConfDepth,
+		MaxBatchSize: cfg.MaxHeaderBatchSize,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		cfg:            cfg,
+		ctx:            ctx,
+		cancel:         cancel,
+		contract:       contract,
+		headerSelector: confirmedHeaderSelector,
+		backend:        cfg.L1Client,
+	}, nil
+}
+
+func (s *Service) Loop(ctx context.Context) {
+	newHeads := make(chan *types.Header, 1000)
+	subscription, err := s.backend.SubscribeNewHead(s.ctx, newHeads)
+	if err != nil {
+		fmt.Print("unable to subscribe to new heads: %v\n", err)
+		s.Stop()
+		return
+	}
+	defer subscription.Unsubscribe()
+
+	start := uint64(0)
+	for {
+		select {
+		case header := <-newHeads:
+			log.Info("Received new header", "header", header.Hash)
+			for {
+				err := s.Update(start, header)
+				if err != nil && err != errNoNewBlocks {
+					fmt.Printf("Unable to update indexer: %v\n", err)
+				}
+				break
+			}
+		case <-s.ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *Service) fetchBlockEventIterator(start, end uint64) (
+	*ctc.CanonicalTransactionChainTransactionEnqueuedIterator, error) {
+
+	const NUM_RETRIES = 5
+	var err error
+	for retry := 0; retry < NUM_RETRIES; retry++ {
+		ctxt, cancel := context.WithTimeout(s.ctx, DefaultConnectionTimeout)
+		defer cancel()
```

Comment:
Avoid `defer` in a loop - can cause leaks since the `defer` will run only after the loop is exited. Instead, call `cancel()` here manually anywhere the loop can exit.

Reply:
Fixed. Thanks, nice call.